Add FXIOS-9682 TabTrayStateTests

yusuphjoluwasen opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-9682)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/21299)

## :bulb: Description
Added Test for TabTrayState
Added unit tests for TabTrayAction with TabTrayActionType:
- didLoadTabTray
- changePanel
- dismissTabTray
- firefoxAccountChanged

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabTrayStateTests.swift:
```diff
@@ -0,0 +1,97 @@
+@testable import Client
+
+final class TabTrayStateTests: XCTestCase {
+    override func setUp() {
```
if `setUp` and `tearDown` are empty methods they can be both deleted

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabTrayStateTests.swift:
```diff
@@ -0,0 +1,97 @@
+    func testDismissTabTrayAction() {
+        let initialState = createSubject()
+        let reducer = tabTrayReducer()
+        XCTAssertEqual(initialState.shouldDismiss, false)
```
It is better to have all the assertions together so we can respect the `Given When Then` pattern. If that is not possible then an option coul'd be to have two tests case

FilippoZazzeroni reviewed (changes requested):
@yusuphjoluwasen thanks a lot for the contribution you did great job 🎉 , only one side thing i see that some test cases are missing since `TabTrayState` is handling more cases, could you please add them. here is a list: 
- didChangePanel
- refreshTabs
- showToast
- closeAllTabs
- defaultState

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabTrayStateTests.swift:
```diff
@@ -0,0 +1,269 @@
+@testable import Client
+
+final class TabTrayStateTests: XCTestCase {
+    
```
Super neat: we can delete this line.

FilippoZazzeroni reviewed (commented):
This is great Thanks so much 🎉  i see only that `defaultState` is missing, i know you added `testInitialState` and it s great but default state it is a different method.

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabTrayStateTests.swift:
```diff
@@ -0,0 +1,302 @@
+        let initialState = createSubject()
+        let reducer = tabTrayReducer()
+        
+        let unknownAction = getUnknownAction()
```
I think this could be much easier IMO, We could test directly:
```
XCTassertEqual(TabTrayState.defaultState(initialState), initialState)
```
the other asserts can be skipped since the state is equatable and it is all evaluated in the first assertion.
Unknow action is not needed IMO we can just test the default method directly avoiding calling the reducer

yusuphjoluwasen replied:
Thank you for your great suggestions. I would like to clarify something: since we currently do not have a static defaultState method in TabTrayState, I have initialized TabTrayState with default values as a variable in TabTrayStateTests. Would you prefer it to be implemented in a different way

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabTrayStateTests.swift:
```diff
@@ -0,0 +1,302 @@
+        return  UnknownMockAction(windowUUID: .XCTestDefaultUUID, actionType: MockActionType())
+    }
+    
+    private class UnknownMockAction: Action {
```
We can delete this

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabTrayStateTests.swift:
```diff
@@ -0,0 +1,302 @@
+        }
+    }
+    
+    private struct MockActionType: ActionType {
```
same

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabTrayStateTests.swift:
```diff
@@ -284,19 +273,18 @@ final class TabTrayStateTests: XCTestCase {
-    private struct MockActionType: ActionType {
-        let description = "Unknown Action"
+    private var createDefaultState:TabTrayState {
+        return TabTrayState(
```
I'm sorry i've not explained my self properly. This is not needed, we'd like to test the method `TabTrayState.defaultState(from state: )`. the defaultState method returns a TabTrayState starting from another TabTrayState but only passing the required parameters to the initializer thus clearing any transient state properties.

This is the current defaultState implementation:
```swift
static func defaultState(from state: TabTrayState) -> TabTrayState {
        return TabTrayState(windowUUID: state.windowUUID,
                            isPrivateMode: state.isPrivateMode,
                            selectedPanel: state.selectedPanel,
                            normalTabsCount: state.normalTabsCount,
                            hasSyncableAccount: state.hasSyncableAccount)
    }
```

Looking to the TabTrayState i was wrong it is not enough to test that the two states are identical because they coul'd be different so i'd suggest something like:

```swift 
func testDefaultState() {
      var initialState = createSubject()
      // those three are transient properties so when getting the default state they should be restored 
      initialState.shouldDismiss = true
      initialState.toastType = .addBookmark
      initialState.showCloseConfirmation = true
      let defaultState = TabTrayState.defaultState(from state: initialState)
      
      XCTassertEqual(defaltState.windowUUID, initialState.windowUUID)
      ... same for the mandatory fields
      XCTassertNotEqual(defaultState.shouldDismiss, intialState.shouldDismiss)
      .. same for all transient fields
}
```
Let me know if it makes sense, thanks again for the great contribution 😃

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabTrayStateTests.swift:
```diff
@@ -284,19 +273,18 @@ final class TabTrayStateTests: XCTestCase {
     }
 }
 
+
```
Nit: this can be deleted

FilippoZazzeroni reviewed (approved):
LGTM thanks so much and great work 🎉 🎉  just a quick hint for next PRs try to avoid force push otherwise i don't see differences between previous commits.

yusuphjoluwasen commented:
Noted, Thanks

FilippoZazzeroni commented:
@yusuphjoluwasen there some lint issues, Do you mind solving them ? thanks so much

FilippoZazzeroni commented:
Hi @yusuphjoluwasen could you please update the PR to solve the swiftlint errors so we can merge it ? Thanks so much 😃

yusuphjoluwasen commented:
Thanks, @FilippoZazzeroni, I have fixed the SwiftLint errors and merged the main branch to prevent any merge conflicts. I also noticed that TabTrayState has been updated, so I went ahead and updated the unit tests accordingly to reflect those changes.

FilippoZazzeroni reviewed (approved):
Perfect thanks so much @yusuphjoluwasen great work 🎉 🎉
